core/drivers/stm32_rtc.c

```diff
@@ -630,9 +787,222 @@ TEE_Result stm32_rtc_driver_is_initialized(void)
+	else
+		io_clrbits32(rtc_base + RTC_CR, RTC_CR_ALRAIE | RTC_CR_ALRAE);
+
+	stm32_rtc_clear_events(RTC_SCR_CALRAF);
+
+	stm32_rtc_write_protect();
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rtc_valid_alarm_time(struct rtc *rtc,
+					     struct optee_rtc_time *tm)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct optee_rtc_time max = { };
+	uint32_t next_month = 0;
+	uint32_t next_year = 0;
+
+	/*
+	 * Assuming current date is M-D-Y H:M:S.
+	 * RTC alarm can't be set on a specific month and year.
+	 * So the valid alarm range is:
+	 *	M-D-Y H:M:S < alarm <= (M+1)-D-Y H:M:S
+	 */
+	res = stm32_rtc_get_time(rtc, &max);
+	if (res)
+		return res;
+
+	/*
+	 * Don't allow alarm to be set in the past.
+	 */
+	if (rtc_timecmp(&max, tm) > 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/*
+	 * Find the next month and the year of the next month.
+	 * Note: tm_mon and next_month are from 0 to 11
+	 */
+	next_month = max.tm_mon + 1;
+	if (next_month == 12) {
+		next_month = 0;
+		next_year = max.tm_year + 1;
+	} else {
+		next_year = max.tm_year;
+	}
+
+	if (max.tm_mday < tm->tm_mday) {
+		/*
+		 * Alarm is next month
+		 * Must be sure it does not exceed the number of days in
+		 * the next month.
+		 */
+		if (tm->tm_mday > rtc_get_month_days(next_month, next_year))
+			return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	max.tm_year = next_year;
+	max.tm_mon = next_month;
+
+	if (rtc_timecmp(&max, tm) < 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rtc_set_alarm(struct rtc *rtc,
+				      struct optee_rtc_alarm *alarm)
+{
+	struct optee_rtc_time *alarm_time = NULL;
+	vaddr_t rtc_base = get_base();
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	uint32_t alrmar = 0;
+	uint32_t cr = io_read32(rtc_base + RTC_CR);
+
+	alarm_time = &alarm->time;
+
+	res = stm32_rtc_valid_alarm_time(rtc, alarm_time);
+	if (res)
+		return res;
+
+	alrmar = 0;
+	/* tm_year and tm_mon are not used because not supported by RTC */
+	alrmar |= ((alarm_time->tm_mday / 10) << RTC_ALRMXR_DATE_TENS_SHIFT) &
+		  RTC_ALRMXR_DATE_TENS_MASK;
+	alrmar |= ((alarm_time->tm_mday % 10) << RTC_ALRMXR_DATE_UNITS_SHIFT) &
+		  RTC_ALRMXR_DATE_UNITS_MASK;
+	/* 24-hour format */
+	alrmar &= ~RTC_ALRMXR_PM;
+	alrmar |= ((alarm_time->tm_hour / 10) << RTC_ALRMXR_HOUR_TENS_SHIFT) &
+		  RTC_ALRMXR_HOUR_TENS_MASK;
+	alrmar |= ((alarm_time->tm_hour % 10) << RTC_ALRMXR_HOUR_UNITS_SHIFT) &
+		  RTC_ALRMXR_HOUR_UNITS_MASK;
+	alrmar |= ((alarm_time->tm_min / 10) << RTC_ALRMXR_MIN_TENS_SHIFT) &
+		  RTC_ALRMXR_MIN_TENS_MASK;
+	alrmar |= ((alarm_time->tm_min % 10) << RTC_ALRMXR_MIN_UNITS_SHIFT) &
+		  RTC_ALRMXR_MIN_UNITS_MASK;
+	alrmar |= ((alarm_time->tm_sec / 10) << RTC_ALRMXR_SEC_TENS_SHIFT) &
+		  RTC_ALRMXR_SEC_TENS_MASK;
+	alrmar |= ((alarm_time->tm_sec % 10) << RTC_ALRMXR_SEC_UNITS_SHIFT) &
+		  RTC_ALRMXR_SEC_UNITS_MASK;
+
+	stm32_rtc_write_unprotect();
+
+	/* Disable Alarm */
+	cr &= ~RTC_CR_ALRAE;
+	io_write32(rtc_base + RTC_CR, cr);
+
```

Comment:
Shouldn't we check here what we're not writing to the same time as the one returned by `stm32_rtc_get_time(rtc, &current_time);` ?

Reply:
`stm32_rtc_valid_alarm_time()` could be called right before `io_write32(rtc_base + RTC_ALRMAR, alrmar);` to reduce the risk of setting an alarm in the past if the alarm was to be set to the next second and the second has incremented in between the call above and the write to the register

Else, maybe we can think of preventing setting an alarm that is too close to the current time. E.g in the current/next second?